onnxruntime/core/providers/rocm/RoctracerLogger.cc

```diff
@@ -0,0 +1,353 @@
+#include "RoctracerLogger.h"
+
+#include <cstring>
+#include <chrono>
+#include <time.h>
+
+//#include "Demangle.h"
+//#include "output_base.h"
+#include "ThreadUtil.h"
+
+typedef uint64_t timestamp_t;
+
+static timestamp_t timespec_to_ns(const timespec& time) {
+    return ((timestamp_t)time.tv_sec * 1000000000) + time.tv_nsec;
+  }
+
+using namespace std::chrono;
+
+constexpr size_t kBufSize(2 * 1024 * 1024);
+
+RoctracerLogger& RoctracerLogger::singleton() {
+  static RoctracerLogger instance;
+  return instance;
+}
+
+RoctracerLogger::RoctracerLogger() {
+  gpuTraceBuffers_ = std::make_unique<std::list<RoctracerActivityBuffer>>();
+}
+
+RoctracerLogger::~RoctracerLogger() {
+  stopLogging();
+  endTracing();
+}
+
+namespace {
+  thread_local std::deque<uint64_t> t_externalIds[RoctracerLogger::CorrelationDomain::size];
+}
+
+void RoctracerLogger::pushCorrelationID(uint64_t id, CorrelationDomain type) {
+  if (!singleton().externalCorrelationEnabled_) {
+    return;
+  }
+  t_externalIds[type].push_back(id);
```

Comment:
>t_externalIds[type].push_back(id); [](http://example.com/codeflow?start=2&length=34)

so for rocm case, the pushCorrelationID(...) will not place an id on device?

Reply:
This is a limitation of libroctracer.  If we have roctracer pass back digest-style api calls (a buffer at a time like cupti does) then those pages lack the call arguments.  So, as you can see, we are collecting the calls and args inline.  Roctracer only generates external correlation events when using that page-at-time method.  This is fine, my inline method is faster, gets the args, and can generate external correlation.

Unless you understand more about external correlation implementation than I do...  your statement about "place an id on device" is a misconception.  All that happens is that a mapping from internal correlation id to (one or more) external correlation ids is generated when the api call is made.  The 'device' is running ops and tracking them by internal correlation id (that is shares with the generating api call), just like normal.

Because I have 'bitten the bullet' and I am doing inline collection I can implement external correlation here.  Without using inline callbacks cupti would have to do this for me and deliver a bunch of records.  My way is clearly faster, so it probably turned into an advantage.